Adding View Components can cause a DB contraint violation

**Describe the bug**

Adding multiple view components, saving between each view component, will cause a constraint violation from the database

```
{
  "details": "Error id bcc68640-37c1-4920-9d19-3f43f079f552-1, org.hibernate.exception.ConstraintViolationException: could not execute statement [ERROR: duplicate key value violates unique constraint \"viewcomponent_view_id_headername_key\"] [insert into viewcomponent (headerName,headerOrder,labels,render,view_id,id) values (?,?,?,?,?,?)]",
  "stack": "org.hibernate.exception.ConstraintViolationException: could not execute statement [ERROR: duplicate key value violates unique constraint \"viewcomponent_view_id_headername_key\"] [insert into viewcomponent (headerName,headerOrder,labels,render,view_id,id) values (?,?,?,?,?,?)]\n\tat org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:95)\n\tat org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:56)\n\tat org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:108)\n\tat org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:278)\n\tat org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.performNonBatchedMutation(AbstractMutationExecutor.java:107)\n\tat org.hibernate.engine.jdbc.mutation.internal.MutationExecutorSingleNonBatched.performNonBatchedOperations(MutationExecutorSingleNonBatched.java:40)\n\tat org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.execute(AbstractMutationExecutor.java:52)\n\tat org.hibernate.persister.entity.mutation.InsertCoordinator.doStaticInserts(InsertCoordinator.java:171)\n\tat org.hibernate.persister.entity.mutation.InsertCoordinator.coordinateInsert(InsertCoordinator.java:112)\n\tat org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:2790)\n\tat org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:102)\n\tat org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:635)\n\tat org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:502)\n\tat org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:364)\n\tat org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)\n\tat org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)\n\tat org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1412)\n\tat org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1398)\n\tat io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.flush(TransactionScopedSession.java:234)\n\tat org.hibernate.engine.spi.SessionLazyDelegator.flush(SessionLazyDelegator.java:76)\n\tat org.hibernate.Session_OpdLahisOZ9nWRPXMsEFQmQU03A_Synthetic_ClientProxy.flush(Unknown Source)\n\tat io.hyperfoil.tools.horreum.svc.UIServiceImpl.doUpdate(UIServiceImpl.java:65)\n\tat io.hyperfoil.tools.horreum.svc.UIServiceImpl.updateView(UIServiceImpl.java:35)\n\tat io.hyperfoil.tools.horreum.svc.UIServiceImpl_Subclass.updateView$$superforward(Unknown Source)\n\tat io.hyperfoil.tools.horreum.svc.UIServiceImpl_Subclass$$function$$1.apply(Unknown Source)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)\n\tat io.hyperfoil.tools.horreum.server.RolesInterceptor.intercept(RolesInterceptor.java:70)\n\tat io.hyperfoil.tools.horreum.server.RolesInterceptor_Bean.intercept(Unknown Source)\n\tat io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)\n\tat io.quarkus.security.runtime.interceptor.SecurityHandler.handle(SecurityHandler.java:47)\n\tat io.quarkus.security.runtime.interceptor.RolesAllowedInterceptor.intercept(RolesAllowedInterceptor.java:29)\n\tat io.quarkus.security.runtime.interceptor.RolesAllowedInterceptor_Bean.intercept(Unknown Source)\n\tat io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:136)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)\n\tat io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)\n\tat io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)\n\tat io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor.intercept(StandardSecurityCheckInterceptor.java:44)\n\tat io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor_RolesAllowedInterceptor_Bean.intercept(Unknown Source)\n\tat io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)\n\tat io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)\n\tat io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)\n\tat io.hyperfoil.tools.horreum.svc.UIServiceImpl_Subclass.updateView(Unknown Source)\n\tat io.hyperfoil.tools.horreum.api.internal.services.UIService$quarkusrestinvoker$updateView_b984046b8fcd1c0136c14c12c9294f9e3713bbd8.invoke(Unknown Source)\n\tat org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)\n\tat io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)\n\tat org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)\n\tat io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)\n\tat org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)\n\tat org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)\n\tat org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)\n\tat org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Thread.java:840)\nCaused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint \"viewcomponent_view_id_headername_key\"\n\tat org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2713)\n\tat org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2401)\n\tat org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)\n\tat org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:498)\n\tat org.postgresql.jdbc.PgStatement.execute(PgStatement.java:415)\n\tat org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190)\n\tat org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:152)\n\tat io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)\n\tat org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:275)\n\t... 60 more"
}
```

**To Reproduce**

Steps to reproduce the behavior.

1. Add a View Component to a View
2. Save
3..Add another View Component
4. Try to Save


What is the version of Horreum ? master @ 24e76e1aaccae197e69caa75eb496ed9eb88bdd5

If you are using a development branch; what is the commit id ?

```
git rev-parse HEAD
24e76e1aaccae197e69caa75eb496ed9eb88bdd5
```

I was able to recreate the issue in the unit tests by changing
io.hyperfoil.tools.horreum.svc.TestServiceTest#testUpdateView
And adding the following to the end of the method
```
      ViewComponent vc2 = new ViewComponent();
      vc2.headerName = "Boobar";
      vc2.labels = JsonNodeFactory.instance.arrayNode().add("value");
      defaultView.components.add(vc2);
      updateView(defaultView);
      TestUtil.eventually(() -> {
         em.clear();
         @SuppressWarnings("unchecked") List<JsonNode> list = em.createNativeQuery(
                         "SELECT value FROM dataset_view WHERE dataset_id = ?1 AND view_id = ?2")
                 .setParameter(1, event.datasetId).setParameter(2, defaultView.id)
                 .unwrap(NativeQuery.class).addScalar("value", JsonBinaryType.INSTANCE)
                 .getResultList();
         return !list.isEmpty() && !list.get(0).isEmpty();
      });
```

This will result in an Exception
```
HTTP Request to /api/ui/view failed, error id: 425a503a-7983-4846-81b1-15a588c42b22-1: org.hibernate.exception.ConstraintViolationException: could not execute statement [ERROR: duplicate key value violates unique constraint "viewcomponent_view_id_headername_key"] [insert into viewcomponent (headerName,headerOrder,labels,render,view_id,id) values (?,?,?,?,?,?)]
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:95)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:56)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:108)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:278)
	at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.performNonBatchedMutation(AbstractMutationExecutor.java:107)
	at org.hibernate.engine.jdbc.mutation.internal.MutationExecutorSingleNonBatched.performNonBatchedOperations(MutationExecutorSingleNonBatched.java:40)
	at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.execute(AbstractMutationExecutor.java:52)
	at org.hibernate.persister.entity.mutation.InsertCoordinator.doStaticInserts(InsertCoordinator.java:171)
	at org.hibernate.persister.entity.mutation.InsertCoordinator.coordinateInsert(InsertCoordinator.java:112)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:2790)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:102)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:635)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:502)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:364)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1412)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1398)
	at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.flush(TransactionScopedSession.java:234)
	at org.hibernate.engine.spi.SessionLazyDelegator.flush(SessionLazyDelegator.java:76)
	at org.hibernate.Session_OpdLahisOZ9nWRPXMsEFQmQU03A_Synthetic_ClientProxy.flush(Unknown Source)
	at io.hyperfoil.tools.horreum.svc.UIServiceImpl.doUpdate(UIServiceImpl.java:62)
	at io.hyperfoil.tools.horreum.svc.UIServiceImpl.updateView(UIServiceImpl.java:36)
	at io.hyperfoil.tools.horreum.svc.UIServiceImpl_Subclass.updateView$$superforward(Unknown Source)
	at io.hyperfoil.tools.horreum.svc.UIServiceImpl_Subclass$$function$$1.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)
	at io.hyperfoil.tools.horreum.server.RolesInterceptor.intercept(RolesInterceptor.java:70)
	at io.hyperfoil.tools.horreum.server.RolesInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)
	at io.quarkus.security.runtime.interceptor.SecurityHandler.handle(SecurityHandler.java:47)
	at io.quarkus.security.runtime.interceptor.RolesAllowedInterceptor.intercept(RolesAllowedInterceptor.java:29)
	at io.quarkus.security.runtime.interceptor.RolesAllowedInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:136)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
	at io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor.intercept(StandardSecurityCheckInterceptor.java:44)
	at io.quarkus.resteasy.reactive.server.runtime.StandardSecurityCheckInterceptor_RolesAllowedInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
	at io.hyperfoil.tools.horreum.svc.UIServiceImpl_Subclass.updateView(Unknown Source)
	at io.hyperfoil.tools.horreum.api.internal.services.UIService$quarkusrestinvoker$updateView_b984046b8fcd1c0136c14c12c9294f9e3713bbd8.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "viewcomponent_view_id_headername_key"
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2713)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2401)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:498)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:415)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:152)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:275)
	... 60 more
```

--- a/docs/site/content/en/openapi/openapi.yaml
+++ b/docs/site/content/en/openapi/openapi.yaml
@@ -1956,6 +1956,43 @@ paths:
           default: true
           type: boolean
         example: false
+      - name: filter
+        in: query
+        description: either a required json sub-document or path expression
+        schema:
+          default: "{}"
+          type: string
+        example:
+          key: requiredValue
+      - name: sort
+        in: query
+        description: label name for sorting
+        schema:
+          default: ""
+          type: string
+      - name: direction
+        in: query
+        description: either Ascending or Descending
+        schema:
+          default: Ascending
+          type: string
+        example: count
+      - name: limit
+        in: query
+        description: the maximum number of results to include
+        schema:
+          format: int32
+          default: 2147483647
+          type: integer
+        example: 10
+      - name: page
+        in: query
+        description: which page to skip to when using a limit
+        schema:
+          format: int32
+          default: 0
+          type: integer
+        example: 2
       responses:
         "200":
           description: OK

--- a/horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/internal/services/UIService.java
+++ b/horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/internal/services/UIService.java
@@ -22,7 +22,7 @@ public interface UIService {
 
     @POST
     @Path("view")
-    int updateView(@RequestBody(required = true) View view);
+    View updateView(@RequestBody(required = true) View view);
 
     @POST
     @Path("views")

--- a/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/UIServiceImpl.java
+++ b/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/UIServiceImpl.java
@@ -28,7 +28,7 @@ public class UIServiceImpl implements UIService {
     @RolesAllowed("tester")
     @WithRoles
     @Transactional
-    public int updateView(View dto) {
+    public View updateView(View dto) {
         if (dto.testId <= 0) {
             throw ServiceException.badRequest("Missing test id on view");
         }
@@ -49,7 +49,7 @@ public class UIServiceImpl implements UIService {
         }
     }
 
-    private int doUpdate(TestDAO test, ViewDAO view) {
+    private View doUpdate(TestDAO test, ViewDAO view) {
         view.ensureLinked();
         view.test = test;
         if (view.id == null || view.id < 0) {
@@ -63,7 +63,7 @@ public class UIServiceImpl implements UIService {
         test.views.add(view);
         test.persist();
         em.flush();
-        return view.id;
+        return ViewMapper.from(view);
     }
 
     @Override

--- a/horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/BaseServiceTest.java
+++ b/horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/BaseServiceTest.java
@@ -388,12 +388,12 @@ public class BaseServiceTest {
               .statusCode(204);
    }
 
-   protected int createView(View view) {
+   protected View createView(View view) {
       return jsonRequest()
               .body(view)
               .post("/api/ui/view")
               .then()
-              .statusCode(200).extract().body().as(Integer.class);
+              .statusCode(200).extract().body().as(View.class);
    }
 
    protected List<View> getViews(int testId) {
@@ -875,7 +875,7 @@ public class BaseServiceTest {
               readFile(p.resolve("roadrunner_view.json").toFile()), View.class);
       assertEquals("Default", view.name);
       view.testId = t.id;
-      view.id = createView(view);
+      view = createView(view);
 
       Schema s = new ObjectMapper().readValue(
               readFile(p.resolve("acme_benchmark_schema.json").toFile()), Schema.class);

--- a/horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/TestServiceTest.java
+++ b/horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/TestServiceTest.java
@@ -167,10 +167,10 @@ public class TestServiceTest extends BaseServiceTest {
    }
 
    private void updateView(View view) {
-      Integer viewId = jsonRequest().body(view).post("/api/ui/view")
-            .then().statusCode(200).extract().body().as(Integer.class);
+      View newView = jsonRequest().body(view).post("/api/ui/view")
+            .then().statusCode(200).extract().body().as(View.class);
       if (view.id != null) {
-         assertEquals(view.id, viewId);
+         assertEquals(view.id, newView.id);
       }
    }
 

--- a/horreum-web/src/api.tsx
+++ b/horreum-web/src/api.tsx
@@ -258,7 +258,7 @@ export function fetchViews(testId: number, alerting: AlertContextType): Promise<
 export function updateAccess(id: number, owner: string, access: Access, alerting: AlertContextType) : Promise<void> {
     return apiCall(testApi.updateAccess(id, access, owner), alerting, "UPDATE_ACCESS", "Failed to update test access");
 }
-export function updateView(alerting: AlertContextType, testId: number, view: View): Promise<number> {
+export function updateView(alerting: AlertContextType, testId: number, view: View): Promise<View> {
     for (const c of view.components) {
         if (c.labels.length === 0) {
             alerting.dispatchError(

--- a/horreum-web/src/domain/tests/Views.tsx
+++ b/horreum-web/src/domain/tests/Views.tsx
@@ -112,7 +112,6 @@ export default function Views({ testId, testOwner, funcsRef, onModified, ...prop
     const [views, setViews] = useState<ViewExtended[]>([])
     const [deleted, setDeleted] = useState<number[]>([])
     const [selectedView, setSelectedView] = useState<ViewExtended>()
-
     useEffect(() => {
         // Perform a deep copy of the view object to prevent modifying store
         const copy = deepCopy(props.views)
@@ -123,21 +122,39 @@ export default function Views({ testId, testOwner, funcsRef, onModified, ...prop
     }, [props.views])
 
     funcsRef.current = {
-        save: () =>
-            Promise.all([
+        save: () =>{
+            return Promise.all([
                 ...views
                     .filter(v => v.modified)
-                    .map(view => updateView(alerting, testId, view).then(id => (view.id = id))),
-                ...deleted.map(id => deleteView(alerting, testId, id)),
-            ]).then(() => {
+                    .map(async view => {
+                        return updateView(alerting, testId, view)
+                    }),
+                ...deleted.map(id => { deleteView(alerting, testId, id); return null})
+            ])
+            //removes the output of the ...deleted.map to satisfy TypeScript type check
+            .then(hasNulls => hasNulls.filter(v => v !=null) as View[])
+            .then((updatedViews) => {
+                //A better solution is for the parent to provide updated views props whenever views change
+                const newViews = [...updatedViews]
+                //add any existing views that were not updated and have valid ids
+                views
+                    .filter(oldView => oldView.id >= 0)//exclude any local temporary views
+                    .forEach(oldView => ( 
+                        newViews.some((newView : ViewExtended) => oldView.id == newView.id)
+                            ? null : newViews.unshift(oldView) //use unshift to preserve view order
+                        ) 
+                    ) 
                 setDeleted([])
-                setViews(
-                    views.map(v => {
-                        delete v.modified
-                        return v
-                    })
-                )
-            }),
+                //have to update selectedView because if it was a new view (id < 0) it will no longer be in views
+                if(selectedView && 'id' in selectedView){
+                    const found = newViews.filter(v => v.id === selectedView.id)
+                    if(found && found.length === 1){
+                        setSelectedView(found[0]);
+                    }
+                }
+                setViews(newViews);
+            })
+        },
         reset: () => {
             setDeleted([])
             setViews(deepCopy(props.views))

